homeassistant/components/mazda/button.py

```diff
@@ -0,0 +1,135 @@
+    """Handle a request to refresh the vehicle status."""
+    await handle_button_press(entity)
+
+    await entity.coordinator.async_refresh()
+
+
+@dataclass
+class MazdaButtonRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    # Suffix to be appended to the vehicle name to obtain the button name
+    name_suffix: str
+
+
+@dataclass
+class MazdaButtonEntityDescription(
+    ButtonEntityDescription, MazdaButtonRequiredKeysMixin
+):
+    """Describes a Mazda button entity."""
+
+    # Function to determine whether the vehicle supports this button, given the coordinator data
+    is_supported: Callable[[dict[str, Any]], bool] = lambda data: True
+
+    async_press: Callable[[MazdaButtonEntity], Awaitable] = handle_button_press
+
+
+BUTTON_ENTITIES = [
+    MazdaButtonEntityDescription(
+        key="start_engine",
+        name_suffix="Start Engine",
+        icon="mdi:engine",
+    ),
+    MazdaButtonEntityDescription(
+        key="stop_engine",
+        name_suffix="Stop Engine",
+        icon="mdi:engine-off",
+    ),
+    MazdaButtonEntityDescription(
+        key="turn_on_hazard_lights",
+        name_suffix="Turn On Hazard Lights",
+        icon="mdi:hazard-lights",
+    ),
+    MazdaButtonEntityDescription(
+        key="turn_off_hazard_lights",
+        name_suffix="Turn Off Hazard Lights",
+        icon="mdi:hazard-lights",
+    ),
+    MazdaButtonEntityDescription(
+        key="start_charging",
+        name_suffix="Start Charging",
+        icon="mdi:ev-station",
+        is_supported=lambda data: data["isElectric"],
+    ),
+    MazdaButtonEntityDescription(
+        key="stop_charging",
+        name_suffix="Stop Charging",
+        icon="mdi:ev-station",
+        is_supported=lambda data: data["isElectric"],
+    ),
+    MazdaButtonEntityDescription(
+        key="refresh_vehicle_status",
+        name_suffix="Refresh Status",
+        icon="mdi:refresh",
+        async_press=handle_refresh_vehicle_status,
+        is_supported=lambda data: data["isElectric"],
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the button platform."""
+    client = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][DATA_COORDINATOR]
+
+    entities: list[ButtonEntity] = []
+
+    for index, data in enumerate(coordinator.data):
+        for description in BUTTON_ENTITIES:
+            if description.is_supported(data):
+                entities.append(
+                    MazdaButtonEntity(client, coordinator, index, description)
+                )
+
+    async_add_entities(entities)
+
+
+class MazdaButtonEntity(MazdaEntity, ButtonEntity):
+    """Representation of a Mazda button."""
+
+    entity_description: MazdaButtonEntityDescription
+
+    def __init__(self, client, coordinator, index, description):
```

Comment:
Type hints?

Reply:
For a future PR, could `client` be made an attribute of `coordinator`?

Reply:
Added type hints for `__init__`.

> For a future PR, could `client` be made an attribute of `coordinator`?

I don't see why not. I'll try to add this in a future PR.